middlewares/retry_test.go

```diff
@@ -0,0 +1,79 @@
+package middlewares
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+func TestRetry(t *testing.T) {
+	testCases := []struct {
+		failAtCalls    []int
+		retryCount     int
+		responseStatus int
+	}{
+		{
+			failAtCalls:    []int{1, 2},
+			retryCount:     3,
+			responseStatus: http.StatusOK,
+		},
+		{
+			failAtCalls:    []int{1, 2},
+			retryCount:     2,
+			responseStatus: http.StatusBadGateway,
+		},
+	}
+
+	for _, tc := range testCases {
+		// bind tc locally
+		tc := tc
+		tcName := fmt.Sprintf("FailAtCalls(%v) RetryCount(%v)", tc.failAtCalls, tc.retryCount)
+
+		t.Run(tcName, func(t *testing.T) {
+			t.Parallel()
+
+			var httpHandler http.Handler
+			httpHandler = &networkFailingHTTPHandler{failAtCalls: tc.failAtCalls}
+			httpHandler = NewRetry(tc.retryCount, httpHandler)
+
+			recorder := httptest.NewRecorder()
+			req, err := http.NewRequest("GET", "http://localhost:3000/ok", ioutil.NopCloser(nil))
+			if err != nil {
+				panic(fmt.Sprintf("NewRequest(): %s", err))
+			}
+
+			httpHandler.ServeHTTP(recorder, req)
+
+			if tc.responseStatus != recorder.Code {
+				t.Errorf("wrong status code, expected %v: got %v", tc.responseStatus, recorder.Code)
+			}
+		})
+	}
+}
+
+// An http.Handler implementation you can use to test retries.
```

Comment:
Should start with the type you are describing, i.e.:

_networkFailingHTTPHandler is an http.Handler implementation you can use to test retries._